src/peft/tuners/poly/model.py

```diff
@@ -0,0 +1,151 @@
+        for name, module in new_module.named_modules():
+            if (self.prefix in name) or ("ranknum" in name):
+                weight = child.qweight if hasattr(child, "qweight") else child.weight
+                module.to(weight.device)
+
+    def _mark_only_adapters_as_trainable(self) -> None:
+        for n, p in self.model.named_parameters():
+            if self.prefix not in n:
+                p.requires_grad = False
+
+    @staticmethod
+    def _create_new_module(poly_config, adapter_name, target, **kwargs):
+        if isinstance(target, BaseTunerLayer):
+            target_base_layer = target.get_base_layer()
+        else:
+            target_base_layer = target
+
+        if isinstance(target_base_layer, torch.nn.Linear):
+            return Linear(target, adapter_name, poly_config, **kwargs)
+        else:
+            raise ValueError(
+                f"Target module {target} is not supported. Currently, only the following modules are supported: "
+                "`torch.nn.Linear`, `torch.nn.Embedding`, `torch.nn.Conv2d`, `transformers.pytorch_utils.Conv1D`."
+            )
+
+    def __getattr__(self, name: str):
+        """Forward missing attributes to the wrapped module."""
+        try:
+            return super().__getattr__(name)  # defer to nn.Module's logic
+        except AttributeError:
+            return getattr(self.model, name)
+
+    def get_peft_config_as_dict(self, inference: bool = False):
+        config_dict = {}
+        for key, value in self.peft_config.items():
+            config = {k: v.value if isinstance(v, Enum) else v for k, v in asdict(value).items()}
+            if inference:
+                config["inference_mode"] = True
+        config_dict[key] = config
+        return config
+
+    def _set_adapter_layers(self, enabled=True):
+        for module in self.model.modules():
+            if isinstance(module, (PolyLayer, ModulesToSaveWrapper)):
+                module.enable_adapters(enabled)
+
+    def enable_adapter_layers(self):
+        self._set_adapter_layers(enabled=True)
+
+    def disable_adapter_layers(self):
+        self._set_adapter_layers(enabled=False)
+
+    def set_adapter(self, adapter_name):
+        for module in self.model.modules():
+            if isinstance(module, PolyLayer):
+                module.set_adapter(adapter_name)
+
+    @staticmethod
+    def _prepare_adapter_config(peft_config, model_config):
+        if peft_config.target_modules is None:
+            if model_config["model_type"] not in TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING:
+                raise ValueError("Please specify `target_modules` in `peft_config`")
+            peft_config.target_modules = set(
+                TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING[model_config["model_type"]]
+            )
+        return peft_config
+
+    @classmethod
+    def _add_task_id_pre_hook(cls, module, args, kwargs):
+        task_ids = kwargs.get("task_ids", None)
```

Comment:
Well, this is not quite what I had in mind. My idea was to get rid of the `.task_id` attribute completely. Instead, the forward hook would be added to `poly.Linear` and would inject the `task_id` argument to the `kwargs` of `poly.Linear`.

I guess this method also works, but WDYT?

Reply:
i was also wondering how to make this work properly and "Pythonically"
```python
# when use `PeftModel.forward`
PeftModel.forward(..., task_ids=...)
PeftModel.base_model.forward(..., task_ids=...)
# we in fact call `PolyModel.forward`
PolyModel.forward(..., task_ids=...)
# so i added the hook here and make `task_ids` as an attribute of `poly.Linear`
```
i am not sure how to inject `task_ids` into certain sub-modules like `poly.Linear` of one model like `PolyModel` as you suggested. sub-modules are somehow chained together? - one's input is another's output.

and most importantly, sometimes the `kwargs` is not given to `Linear` layers in`forward` (see Llama below), so it won't be passed into any pre-foward hook? i guess adding a hook here for `{q,k,v}_proj` would not help?  🤔
https://github.com/huggingface/transformers/blob/514de24abfd4416aeba6a6455ad5920f57f3567d/src/transformers/models/llama/modeling_llama.py#L366C13-L366C54
```python
query_states = self.q_proj(hidden_states)
key_states = self.k_proj(hidden_states)
value_states = self.v_proj(hidden_states)
```
also the doc of [register_forward_pre_hook](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_forward_pre_hook)
> If `with_kwargs` is true, the forward pre-hook will be passed **the kwargs given to the forward function**.

Reply:
aha, i managed to make this work as you suggested 😄

using `task_ids` as `kwargs` instead of attribute would indeed be better.

Reply:
hi @BenjaminBossan, I want to ensure I’ve understood your suggestion correctly. Are you proposing that we assign one `hook_fn` to each `poly.Linear` within our `PolyModel` like this?

```python
class PolyModel(BaseTuner):
    def __init__(self, model, config, adapter_name) -> None:
        super().__init__(model, config, adapter_name)
        self.hook_handles = dict()
        self._add_hooks()

    def _add_hooks(self):
        for name, module in self.model.named_modules():
            if isinstance(module, PolyLayer) and isinstance(module, nn.Module):
                self.hook_handles[name] = module.register_forward_pre_hook(
                    **hook_fn**, with_kwargs=True)
    ...

def hook_fn(module, args, kwargs):
    # inject task_ids into kwargs, but how?
    return args, kwargs
```

I’m trying to figure out how to implement `hook_fn` such that it can access and inject `task_ids`, which are provided to the model during the `PolyModel.forward` call.

One approach I thought of is defining `hook_fn` during the forward pass of `PolyModel`. However, this seems problematic in a multi-GPU training scenario, as it would result in multiple hooks being added to the same layer, which is what currently happens in the code [here](https://github.com/TaoSunVoyage/peft/blob/1a7dab4ec3e447d1de8d6af1d24b2fd17400e666/src/peft/tuners/poly/model.py#L139)

Reply:
> However, this seems problematic in a multi-GPU training scenario, as it would result in multiple hooks (one per gpu) being added to the same layer

Good point. Ideally, we could do an atomic "check if hook is present and only add it if not" operation, but I don't think that's possible. Regardless though, I think this should work correctly if we add the hook during `forward`, as that is already device-specific, i.e. after the inputs have been split. I ran a dummy script with DDP and accelerate, adding a forward hook to the child module during the `forward` of the parent module, and it works:

```python
class ChildModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.lin0 = nn.Linear(10, 20)
        self.relu = nn.ReLU()
        self.lin1 = nn.Linear(20, 2)

    def forward(self, X, task_ids=None):
        idx = accelerator.local_process_index
        msg = f"{idx}: {task_ids[0]}"
        print(msg)
        X = X.float()
        X = self.lin0(X)
        X = self.relu(X)
        X = self.lin1(X)
        return X

class ParentModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.mod = ChildModule()

    def forward(self, X, **kwargs):
        def pre_hook(module, args, kwargs):
            X, *_ = args
            task_ids = -100 * X
            kwargs["task_ids"] = task_ids
            return (X,), kwargs

        handle = self.mod.register_forward_pre_hook(pre_hook, with_kwargs=True)

        try:
            result = self.mod(X, **kwargs)
        finally:
            handle.remove()
        return result
```

In the child's `forward`, I print the `task_ids` and they are distinct for each process index, so I think there is no danger of different processes overriding one another's hooks.